tests/std/tests/P1222R4_flat_set/test.cpp

```diff
@@ -408,14 +395,73 @@ void test_non_static_comparer() {
     assert_all_requirements_and_equals(a, {9, 7, 5, -1});
 }
 
+template <template <class...> class Set>
+void test_extract_1() {
+    // Test that the container will be emptied, even if the container's move ctor exits via an exception.
 
-template <class C>
-void test_extract() {
-    constexpr int elements[]{1, 2, 3, 4};
-    C fs{1, 2, 3, 4};
-    auto cont = std::move(fs).extract();
-    assert(fs.empty());
-    assert(ranges::equal(cont, elements));
+    static bool will_throw = false;
+
+    class test_exception : public vector<int> {
+        using base = vector<int>;
+
+    public:
+        using base::base;
+        test_exception(const test_exception&)            = default;
+        test_exception& operator=(const test_exception&) = default;
+        test_exception& operator=(test_exception&&)      = default;
+
+        test_exception(test_exception&& other) : base(static_cast<base&&>(other)) {
+            if (will_throw) {
+                throw 0; // will be caught by "catch (...)"
+            }
+        }
+    };
+
+    will_throw = false;
+    Set<int, std::less<int>, test_exception> fs{4, 3, 2, 1};
+    assert_all_requirements_and_equals(fs, {1, 2, 3, 4});
+    auto extr = std::move(fs).extract();
+    assert(ranges::equal(extr, vector{1, 2, 3, 4}));
+    assert_all_requirements_and_equals(fs, {}); // assert empty
+
+    fs = {4, 3, 2, 1};
+    assert_all_requirements_and_equals(fs, {1, 2, 3, 4});
+    try {
+        will_throw = true;
+        (void) std::move(fs).extract();
+    } catch (...) {
+        will_throw = false;
+        assert_all_requirements_and_equals(fs, {}); // assert empty
+        return;
+    }
+
+    assert(false);
+}
+
+template <template <class...> class Set>
+void test_extract_2() {
+    // Test that the container will be emptied, even if the container's move ctor doesn't empty the container.
+
+    class always_copy : public vector<int> {
+        using base = vector<int>;
+
+    public:
+        using base::base;
+        always_copy(const always_copy&)            = default;
+        always_copy& operator=(const always_copy&) = default;
+
+        // the move ctor & assignment will not empty the container.

```

Comment:
Because `&` has a special meaning, we generally avoid using it in comments. We should spell out "and" here.